lib/auth/grpcserver.go

```diff
@@ -3033,6 +3096,16 @@ func (g *GRPCServer) UpsertNode(ctx context.Context, node *types.ServerV2) (*typ
 	}
 	node.SetAddr(utils.ReplaceLocalhost(node.GetAddr(), p.Addr.String()))
 
+	// Use labels from discovered resources instead of ones reported by discovered ec2 instances
+	filteredNode, err := auth.ServerWithRoles.FilterEC2Labels(ctx, node)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	node, ok = filteredNode.(*types.ServerV2)
+	if !ok {
+		return nil, trace.BadParameter("unexpected type %T", filteredNode)
+	}
+
 	keepAlive, err := auth.ServerWithRoles.UpsertNode(ctx, node)
 	if err != nil {
 		return nil, trace.Wrap(err)
```

Comment:
This will have no effect in most cases.  `UpsertNode` is only called by the node itself if it doesn't have a healthy controls stream.  When healthy, its heartbeats are injected by the auth server directly:
https://github.com/gravitational/teleport/blob/70833b6b3f299d11b7a2a935781b57322c4f69cb/lib/inventory/controller.go#L462

If you want to always intercept a node heartbeat, this should be happening at the level of `auth.Server`, since that will affect all calls to `UpsertNode`, not just calls that get routed through the GRPC server.